### PR TITLE
tree2: Better organize type tests and add recursive schema builder type tests

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Better organize type tests and add recursive schema builder type tests

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/dds/tree2/src/test/feature-libraries/schemaBuilder.spec.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		type _1 = requireTrue<
+			areSafelyAssignable<
+				typeof recursiveStruct,
+				ReturnType<typeof recursiveStruct.structFieldsObject.foo.allowedTypes[0]>
+			>
+		>;
+	});
+
+	it("recursive without special functions2", () => {
+		// This function helps the TypeScript compiler imagine a world where it solves for types in a different order, and thus handles the cases we need.
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/feature-libraries/schemaBuilder.spec.ts:57]: Nit: how does it do this? Any docs / issues we can point to?

File: experimental/dds/tree2/src/test/feature-libraries/typedSchema/typedTreeSchema.spec.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		type _2 = requireFalse<isAssignableTo<typeof jsonBoolean, FieldNodeSchema>>;
+		type _3 = requireFalse<isAssignableTo<typeof jsonBoolean, MapSchema>>;
+		type _4 = requireFalse<isAssignableTo<typeof jsonBoolean, StructSchema>>;
+	}
+
+	{
+		type _1 = requireFalse<isAssignableTo<typeof jsonArray, LeafSchema>>;
+		type _2 = requireTrue<isAssignableTo<typeof jsonArray, FieldNodeSchema>>;
+		type _3 = requireFalse<isAssignableTo<typeof jsonArray, MapSchema>>;
+		// TODO: Fix
```
Comment (Josmithr) [inline experimental/dds/tree2/src/test/feature-libraries/typedSchema/typedTreeSchema.spec.ts:86]: Was this intended to be fixed in this PR? Or in the future?

**Review by Josmithr — APPROVED** [review]

File: experimental/dds/tree2/src/test/feature-libraries/schemaBuilder.spec.ts (hunk 1 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/feature-libraries/schemaBuilder.spec.ts:57]: same issue linked above. As for how: it happens to work. TypeScript does supports some undefined subset of recursive types and this just happes by chance to make it work for reasons that are unspecified and impleemtnation details.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/feature-libraries/schemaBuilder.spec.ts:58]: I have added some more comments.

File: experimental/dds/tree2/src/test/feature-libraries/typedSchema/typedTreeSchema.spec.ts (hunk 2 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/feature-libraries/typedSchema/typedTreeSchema.spec.ts:86]: No, we arn't going to fix that until we finish migrating off the old editable-tree then refactor the schema types.

**Pull request merged**